src/Query/QueryBuilder.php

```diff
@@ -309,14 +309,51 @@ public function fetchFirstColumn(): array
      */
     public function executeQuery(): Result
     {
+        [$params, $types] = $this->boundParameters();
+
         return $this->connection->executeQuery(
             $this->getSQL(),
-            $this->params,
-            $this->types,
+            $params,
+            $types,
             $this->resultCacheProfile,
         );
     }
 
+    /**
+     * Retrieve parameters and types bound to all queries (optional CTEs and main query).
+     *
+     * @return array{
+     *     list<mixed>|array<string, mixed>,
+     *     WrapperParameterTypeArray,
+     * } The parameters and types bound to the CTE queries merged with those bound to the main query.
+     */
+    private function boundParameters(): array
+    {
+        if (count($this->withParts) === 0) {
+            return [$this->params, $this->types];
+        }
```

Comment:
Is this condition really necessary?

Reply:
It avoid merging an empty array if all of the CTE are string (not QueryBuilders)

Reply:
Isn't it done here?

https://github.com/doctrine/dbal/blob/c7e00a7a9c81fbdb73275aadb3b2193fb28bf68c/src/Query/QueryBuilder.php#L348-L350

If I remove the first condition, all tests still pass.

Reply:
Agree, I removed the first condition